proof/access-control/AARCH64/ArchInterrupt_AC.thy

```diff
@@ -0,0 +1,111 @@
+(*
+ * Copyright 2020, Data61, CSIRO (ABN 41 687 119 230)
+ *
+ * SPDX-License-Identifier: GPL-2.0-only
+ *)
+
+theory ArchInterrupt_AC
+imports
+  Interrupt_AC
+begin
+
+context Arch begin global_naming AARCH64
+
+named_theorems Interrupt_AC_assms
+
+definition arch_authorised_irq_ctl_inv ::
+  "'a PAS \<Rightarrow> Invocations_A.arch_irq_control_invocation \<Rightarrow> bool" where
+  "arch_authorised_irq_ctl_inv aag cinv \<equiv>
+     case cinv of (ARMIRQControlInvocation irq x1 x2 trigger) \<Rightarrow>
+       is_subject aag (fst x1) \<and> is_subject aag (fst x2) \<and>
+       (pasSubject aag, Control, pasIRQAbs aag irq) \<in> pasPolicy aag"
+
+lemma arch_invoke_irq_control_pas_refined[Interrupt_AC_assms]:
+  "\<lbrace>pas_refined aag and valid_mdb and arch_irq_control_inv_valid irq_ctl_inv
+                    and K (arch_authorised_irq_ctl_inv aag irq_ctl_inv)\<rbrace>
+   arch_invoke_irq_control irq_ctl_inv
+   \<lbrace>\<lambda>_. pas_refined aag\<rbrace>"
+  apply (cases irq_ctl_inv; simp)
+  apply (wpsimp wp: cap_insert_pas_refined_not_transferable)
+  apply (clarsimp simp: cte_wp_at_caps_of_state clas_no_asid cap_links_irq_def
+                        arch_authorised_irq_ctl_inv_def aag_cap_auth_def
+                        arch_irq_control_inv_valid_def)
+  done
+
+lemma arch_invoke_irq_handler_pas_refined[Interrupt_AC_assms]:
+  "\<lbrace>pas_refined aag and invs and (\<lambda>s. interrupt_states s x1 \<noteq> IRQInactive)\<rbrace>
+   arch_invoke_irq_handler (ACKIrq x1)
+   \<lbrace>\<lambda>_. pas_refined aag\<rbrace>"
+  by wpsimp
+
+lemma arch_invoke_irq_control_respects[Interrupt_AC_assms]:
+  "\<lbrace>integrity aag X st and pas_refined aag and K (arch_authorised_irq_ctl_inv aag acinv)\<rbrace>
+   arch_invoke_irq_control acinv
+   \<lbrace>\<lambda>_. integrity aag X st\<rbrace>"
+  apply (case_tac acinv, clarsimp simp add: setIRQTrigger_def arch_authorised_irq_ctl_inv_def)
+  apply (wpsimp wp: cap_insert_integrity_autarch aag_Control_into_owns_irq
+                    dmo_mol_respects do_machine_op_pas_refined)
+  done
+
+lemma integrity_irq_masks [iff]:
+  "integrity aag X st (s\<lparr>machine_state := machine_state s \<lparr>irq_masks := v\<rparr>\<rparr>) =
+   integrity aag X st s"
+  unfolding integrity_def by simp
+
+lemma arch_invoke_irq_handler_respects[Interrupt_AC_assms]:
+  "\<lbrace>integrity aag X st and pas_refined aag and einvs\<rbrace>
+   arch_invoke_irq_handler (ACKIrq x1)
+   \<lbrace>\<lambda>_. integrity aag X st\<rbrace>"
+  by (wpsimp wp: dmo_wp mol_respects simp: maskInterrupt_def plic_complete_claim_def)
+
+crunch arch_check_irq for inv[Interrupt_AC_assms, wp]: P
+
+end
+
+
+context begin interpretation Arch .
+
+requalify_consts arch_authorised_irq_ctl_inv
```

Comment:
Not for this PR, but there's tech to do this more efficiently, documented in `docs/arch-split.md` (arch_requalify_facts)